daemon/container_linux.go

```diff
@@ -737,27 +737,89 @@ func (container *Container) buildCreateEndpointOptions() ([]libnetwork.EndpointO
 	return createOptions, nil
 }
 
+func parseService(controller libnetwork.NetworkController, service string) (string, string, string) {
+	dn := controller.Config().Daemon.DefaultNetwork
+	dd := controller.Config().Daemon.DefaultDriver
+
+	snd := strings.Split(service, ".")
+	if len(snd) > 2 {
+		return strings.Join(snd[:len(snd)-2], "."), snd[len(snd)-2], snd[len(snd)-1]
+	}
+	if len(snd) > 1 {
+		return snd[0], snd[1], dd
+	}
+	return snd[0], dn, dd
+}
+
+func createNetwork(controller libnetwork.NetworkController, dnet string, driver string) (libnetwork.Network, error) {
+	createOptions := []libnetwork.NetworkOption{}
+	genericOption := options.Generic{}
+
+	// Bridge driver is special due to legacy reasons
+	if runconfig.NetworkMode(driver).IsBridge() {
+		genericOption[netlabel.GenericData] = map[string]interface{}{
+			"BridgeName":            dnet,
+			"AllowNonDefaultBridge": "true",
+		}
+		networkOption := libnetwork.NetworkOptionGeneric(genericOption)
+		createOptions = append(createOptions, networkOption)
+	}
+
+	return controller.NewNetwork(driver, dnet, createOptions...)
+}
+
 func (container *Container) AllocateNetwork() error {
 	mode := container.hostConfig.NetworkMode
+	controller := container.daemon.netController
 	if container.Config.NetworkDisabled || mode.IsContainer() {
 		return nil
 	}
 
-	var err error
+	var networkDriver string
+	service := container.Config.PublishService
+	networkName := mode.NetworkName()
+	canCreateNetwork := false
+	if mode.IsDefault() {
+		canCreateNetwork = true
+		if service != "" {
+			service, networkName, networkDriver = parseService(controller, service)
```

Comment:
If I read this and `parseService` above accurately, a user must supply a "service name" to be able to join a network at `docker run` time. This seems like a rather befuddling conflation of networks and service discovery. What if a user wants a load of containers that don't need to have service names?

I don't see what was the problem with keeping networking and service discovery distinct (and supporting, say, `docker --net=foo ..`). It doesn't make sense to hide networking under service discovery, since users may not always want the latter.